#**MEMÓRIA CACHE - ATIVIDADE 5**

##Alunos:
* Pablo Durkheim Fernandes do Nascimento
* Rafael Ribeiro Franco

Esta tarefa consiste em construir uma simulação de um sistema computacional com cache única 
entre processador e memória principal para testar várias configurações de projetos de cache e, com 
isso, obter as taxas de acertos e falhas existentes.

A simulação deve ser escrita em linguagem Python, utilizando a biblioteca pyCacheSim, como 
mencionada na “Atividade 1”. Para isto, o sistema hipotético que deve ser considerado possui uma 
memória principal de 64K bytes de tamanho, endereçada a byte (ou seja, o tamanho da palavra 
desse sistema é igual a 1 byte). Esse sistema é de um único processador, com uma cache simples de 
dados (cache única), de 4K bytes de tamanho.

Considere também a existência de duas sequências de referências para leitura da memória que estão
listadas em dois arquivos diferentes (“referencia1.txt” e “referencia2.txt”, presentes no SIGAA). 
Para cada uma dessa sequência (também chamada de traço da memória), o seu programa deve 
simular:
> (**a**) Uma cache diretamente mapeada, com blocos de 8, 16, 32 e 64 bytes, respectivamente; e

> (**b**) Uma cache associativa em conjunto com 2, 4, 8 e 16 vias, sendo que cada uma delas 
apresenta blocos de 8, 16, 32 e 64 bytes, respectivamente.

No caso da simulação com cache associativa em conjunto, devem ser testados as três políticas de 
substituição de linha (LRU, FIFO e aleatória).

Para testarmos uma cache totalmente associativa, vamos considerar uma configuração diferente: 
cache única de dados, de 1K bytes de tamanho, com linhas armazenando 8, 16, 32 e 64 bytes, 
respectivamente.

Vê-se, portanto, que deverão ser realizadas 4 (mapeamento direto) + 16 (associativo em conjunto) +
4 (totalmente associativo) simulações diferentes. Em cada uma delas, obtenha as taxas de acerto e 
de falha de cache.

No final, elabore um relatório em que sejam apresentados:

* Uma explicação geral sobre o programa elaborado e a biblioteca, apresentando as vantagens 
percebidas ou as dificuldades encontradas na escrita do código;
* A interpretação do endereço da memória principal em seus vários campos constituíntes para 
cada simulação de cache realizada; 
* O total de erros e acertos em cada uma das simulações; e
* Considerações finais, comparando os desempenhos retornados pelas (i) simulações com 
mapeamento direto e associativo em conjunto, e (ii) simulações com o mapeamento 
totalmente associativo

In [1]:
!pip install pycachesim

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
  Preparing metadata (setup.py) ... done
  Created wheel for pycachesim: filename=pycachesim-0.3.1-cp310-cp310-linux_x86_64.whl size=68949 sha256=029e1a881626caaa7a130380ce814348cc3d8091adab9959e5b54fa64d6d4a7d
  Stored in directory: /root/.cache/pip/wheels/db/d2/12/8228e89cd2f2a1ff8a184440fd7105b775d8f245f576c80408
Successfully built pycachesim


In [47]:
from cachesim import CacheSimulator, Cache, MainMemory

def run_simulation(cache_size, block_size, associativity, reference_file, replacement_policy=None):
    mem = MainMemory(64 * 1024)  # 64K bytes de memória principal
    cache = Cache("Cache", cache_size, associativity, block_size)

    if(replacement_policy!=None):
      cache = Cache("Cache", cache_size, associativity, block_size, replacement_policy)
    mem.load_to(cache)
    mem.store_from(cache)
    cs = CacheSimulator(cache, mem)

    with open(reference_file, "r") as file:
        for line in file:
            address = int(line.strip())
            cs.load(address)
           

    cs.print_stats()

def main():
    cache_size = 4 * 1024  # 4K bytes cache
    block_sizes = [8, 16, 32, 64]
    associativities = [2, 4, 8, 16]
    reference_files = ["referencia1.txt", "referencia2.txt"]
    replacement_policies = ["LRU", "FIFO", "RR"]

    # Simulate diretamente mapeada cache
    print("- - - DIRETAMENTE MAPEADA - - - \n")
    for block_size in block_sizes:
        for reference_file in reference_files:
            print(f"Simulating directly-mapped cache - Block Size: {block_size} bytes - Reference File: {reference_file}")
            run_simulation(cache_size, block_size, 1, reference_file)
            print("\n")

    print("- - - ASSOCIATIVA EM CONJUNTO - - - \n")
    # Simulate cache associativa em conjunto
    for associativity in associativities:
        for block_size in block_sizes:
            for reference_file in reference_files:
                for replacement_policy in replacement_policies:
                    print(f"Simulating set-associative cache - Associativity: {associativity} - Block Size: {block_size} bytes - Reference File: {reference_file} - Replacement Policy: {replacement_policy}")
                    run_simulation(cache_size, block_size, associativity, reference_file, replacement_policy)
                    print("\n")

    print("- - - TOTALMEMNTE ASSOCIATIVA - - - \n")
    # Simulate cache totalmente associativa
    for block_size in block_sizes:
        for reference_file in reference_files:
            print(f"Simulating fully-associative cache - Block Size: {block_size} bytes - Reference File: {reference_file}")
            run_simulation(1024, block_size, 1024 // block_size, reference_file)
            print("\n")

if __name__ == "__main__":
    main()


- - - DIRETAMENTE MAPEADA - - - 

Simulating directly-mapped cache - Block Size: 8 bytes - Reference File: referencia1.txt
CACHE *******HIT******** *******MISS******* *******LOAD******* ******STORE******* ******EVICT*******
Cache  19686 (   19686B)   2814 (    2814B)  22500 (   22500B)      0 (       0B)      0 (       0B)
65536   2814 (    2814B)      0 (       0B)   2814 (    2814B)      0 (       0B)      0 (       0B)


Simulating directly-mapped cache - Block Size: 8 bytes - Reference File: referencia2.txt
CACHE *******HIT******** *******MISS******* *******LOAD******* ******STORE******* ******EVICT*******
Cache  19686 (   19686B)   2814 (    2814B)  22500 (   22500B)      0 (       0B)      0 (       0B)
65536   2814 (    2814B)      0 (       0B)   2814 (    2814B)      0 (       0B)      0 (       0B)


Simulating directly-mapped cache - Block Size: 16 bytes - Reference File: referencia1.txt
CACHE *******HIT******** *******MISS******* *******LOAD******* ******STORE******* ******

#**RESOLUÇÃO:**

Para solucionarmos esta tarefa, utilizamos a biblioteca pyCacheSim, como mencionado no arquivo da atividade 5 da disciplina de Arquitetura de Computadores.

##**Código**: 

O código é dividido em três partes principais:

1. A primeira parte é a importação da biblioteca pycachesim e a definição de três classes: CacheSimulator, Cache e MainMemory. Essas classes são usadas para modelar os componentes da memória cache, como o simulador de cache, a cache em si e a memória principal.

2. A função run_simulation é definida para executar a simulação de um determinado cenário de cache. Ela recebe vários parâmetros, como o tamanho da cache, o tamanho do bloco, a associatividade, o arquivo de referência e a política de substituição (opcional). Essa função carrega o arquivo de referência, executa as operações de leitura e gravação na cache e, no final, imprime as estatísticas da simulação.

3. A função principal main é definida para executar as simulações. Ela define os parâmetros iniciais, como o tamanho da cache, os tamanhos de bloco, as associatividades, os arquivos de referência e as políticas de substituição. Em seguida, ela realiza três tipos de simulações: diretamente mapeado, associativo por conjunto e totalmente associativo. Para cada tipo de cache, ela varia os parâmetros de tamanho do bloco, associatividade, arquivo de referência e política de substituição (no caso do cache associativo por conjunto). Durante a simulação, as informações relevantes são impressas na saída.

No final, a função main é chamada para iniciar a execução das simulações.

**Vantagens e desvantagens:** A utilizaão do código foi útil para a verificação e visualização de muitos testes, contudo, a biblioteca  pyCacheSim não tem muitos recursos ainda, e por isso, obtivemos dificuldades na parte de mapeamento direto e na parte de acessar um dado diretamente (cs.print_stats()).

##**Saída (tabela)**:

A saída gerada pelo código "cs.print_stats()" mostra estatísticas sobre a simulação da cache.

- "HIT": O número de acessos à cache que resultaram em um acerto (hit).
- "MISS": O número de acessos à cache que resultaram em uma falha (miss), ou seja, o dado não estava presente na cache e precisou ser buscado na memória principal.
- "LOAD": O número total de acessos à cache, incluindo tanto os hits quanto os misses.
- "STORE": O número de acessos de escrita na cache.
- "EVICT": O número de blocos que foram evictados (removidos) da cache.

A tabela apresenta os valores tanto para a cache quanto para a memória principal. Na linha "Cache", os números indicam as estatísticas relacionadas à cache em si, enquanto na linha "65536", os números correspondem às estatísticas da memória principal.

Os valores entre parênteses após os números representam o tamanho em bytes dos dados. Por exemplo, "19686 (19686B)" indica que foram acessados 19.686 bytes na cache e "2814 (2814B)" indica que foram acessados 2.814 bytes na memória principal.

**MAPEAMENTO DIRETO:** No mapemaneto direto, percebemos que nº de acertos na parte de Cache aumenta, e o nº de erros cai durante os testes com os arquivos referência 1 e 2, o load permanece o mesmo (22500). **A taxa de acertos e de erros pode ser calculada da seguinte forma: nº de hits na tabela / nº de load.** Sendo assim, para os 1º casos temos da tabela, temos:

> Taxa de acertos Cache: 19686 / 22500 = 0.8749 = 87.49%
> Taxa de erro Cache: 2814 / 22500 = 0.1251 = 12.51%

> Taxa de acertos Cache: 21093 / 22500 = 0.9375 = 93.75%
> Taxa de erro Cache: 1407 / 22500 = 0.0625 = 06.25%

Podemos perceber que a taxa de acerto continua aumentando ao longo dos testes.

**ASSOCIATIVA EM CONJUNTO E TOTALMENTE ASSOCIATIVO:** temos uma semelhança nos dados, visto que, os dados de acertos e erros se repitiram para os casos de LRU, FIFO e RR, o que na nossa avaliação não deveria acontecer.

##**NO GERAL:** o mapaemento direto ganha em velocidade e simplicidade, já o associativa em conjunto é mais lento e possui política de substituição.